### PR TITLE
Add `core::fmt::WriteCursor` for formatting into a borrowed buffer.

### PR DESCRIPTION
**Pull request by jmillikin**
Proposed implementation for [ACP 278](https://github.com/rust-lang/libs-team/issues/278)

### CODE REVIEW

File: library/core/src/fmt/mod.rs [hunk 1]
```diff
@@ -231,6 +231,133 @@ impl<W: Write + ?Sized> Write for &mut W {
...
+            .field("written", &self.written())
+            .finish()
+    }
+}
+
+impl<'a> WriteCursor<'a> {
+    /// Creates a new cursor wrapping the provided buffer.
+    #[unstable(feature = "fmt_write_cursor", issue = "none")]
+    pub fn new(buf: &'a mut [u8]) -> WriteCursor<'a> {
+        // SAFETY: [T] and [MaybeUninit<T>] have the same layout
```
Comment (programmerjake) [inline library/core/src/fmt/mod.rs:275]:
```suggestion
        // SAFETY: [T] and [MaybeUninit<T>] have the same layout and we never write uninit to any bytes
```

File: library/core/src/fmt/mod.rs [hunk 2]
```diff
@@ -231,6 +231,133 @@ impl<W: Write + ?Sized> Write for &mut W {
...
+    #[unstable(feature = "fmt_write_cursor", issue = "none")]
+    pub fn as_str(&self) -> &str {
+        // SAFETY: The buffer is only initialized by copying from `str` values.
+        unsafe { str::from_utf8_unchecked(self.as_bytes()) }
+    }
+
+    /// Consumes the cursor and returns the written portion of the cursor's
+    /// underlying buffer as a `&str`.
+    #[unstable(feature = "fmt_write_cursor", issue = "none")]
+    pub fn into_str(self) -> &'a str {
```
Comment (programmerjake) [inline library/core/src/fmt/mod.rs:331]: maybe the return type should be `&'a mut str`?

File: library/core/src/fmt/mod.rs [hunk 3]
```diff
@@ -231,6 +231,133 @@ impl<W: Write + ?Sized> Write for &mut W {
...
+        // updates `utf8_len`.
+        unsafe { mem::MaybeUninit::slice_assume_init_ref(utf8) }
+    }
+
+    /// Consumes the cursor and returns the written portion of the cursor's
+    /// underlying buffer as a byte slice.
+    ///
+    /// The returned slice contains valid UTF-8.
+    #[unstable(feature = "fmt_write_cursor", issue = "none")]
+    pub fn into_bytes(self) -> &'a [u8] {
```
Comment (programmerjake) [inline library/core/src/fmt/mod.rs:314]: maybe the return type should be `&'a mut [u8]`?

File: library/core/src/fmt/mod.rs (hunk 1 above)
Comment (jmillikin) [reply library/core/src/fmt/mod.rs]: Done.

File: library/core/src/fmt/mod.rs (hunk 3 above)
Comment (jmillikin) [reply library/core/src/fmt/mod.rs]: Great idea! Done.

File: library/core/src/fmt/mod.rs (hunk 2 above)
Comment (jmillikin) [reply library/core/src/fmt/mod.rs]: Done.

**Comment by SUPERCILEX** [comment]
I don't think this is a good idea as it can be solved in a generalized way: https://github.com/rust-lang/libs-team/issues/278#issuecomment-1848778352

**Comment by joshtriplett** [comment]
r? libs-api

**Pull request closed without merging**